### PR TITLE
TF-78 Mark as email read/unread

### PR DESCRIPTION
**Pull request by dab246**
- [x] Open mail -> read
- [x] Action in ThreadView -> read/unread
- [x] EmailView -> click to icon to unread

https://user-images.githubusercontent.com/80730648/133536687-8b5548e8-c376-42ac-9838-e7c927bc2d40.mov

### CODE REVIEW

File: lib/features/email/data/datasource/email_datasource.dart [hunk 1]
```diff
@@ -6,4 +6,6 @@ abstract class EmailDataSource {
   Future<Email> getEmailContent(AccountId accountId, EmailId emailId);
 
   Future<bool> sendEmail(AccountId accountId, EmailRequest emailRequest);
+
+  Future<bool> markAsRead(AccountId accountId, EmailId emailId, bool unread);
```
Comment (hoangdat) [inline lib/features/email/data/datasource/email_datasource.dart:10]: IMO, please create enum:
```
enum ReadActions {
  markAsRead,
  markAsUnread
}
```

File: lib/features/email/presentation/email_controller.dart [hunk 2]
```diff
@@ -73,8 +83,23 @@ class EmailController extends BaseController {
...
+              if (success is MarkAsEmailReadSuccess) {
+                emailActionCurrent = MarkAsEmailReadAction(presentationEmail.id, mailboxCurrent);
+                _updateMailboxDashBoardActionForDesktop(emailActionCurrent);
+              }
+            })
+          );
+    }
+  }
+
+  void _updateMailboxDashBoardActionForDesktop(AppAction? appAction) {
```
Comment (hoangdat) [inline lib/features/email/presentation/email_controller.dart]: at the moment, please update with refresh for MailboxView and ThreadView, in the next sprint we will improve with `/changes`. Thanks

File: lib/features/email/presentation/email_controller.dart [hunk 3]
```diff
@@ -67,13 +80,41 @@ class EmailController extends BaseController {
...
+                }},
+              (success) {
+                if (success is MarkAsEmailReadSuccess) {
+                  emailActionCurrent = MarkAsEmailReadAndUnreadAction();
+                  _updateMailboxDashBoardActionForDesktop(context, emailActionCurrent);
+                }
+                if (readActions == ReadActions.markAsUnread) {
+                  backToThreadView(context);
+                }
+              }));
```
Comment (hoangdat) [inline lib/features/email/presentation/email_controller.dart:101]: quite hard to understand, it is break our agreement on unidirectional data flow. `viewState` is not update correctly base on the change of the new state. Please take care the `viewState`.

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart [hunk 4]
```diff
@@ -19,6 +20,7 @@ class MailboxDashBoardController extends BaseController {
   final selectedEmail = Rxn<PresentationEmail>();
   final accountId = Rxn<AccountId>();
   final userProfile = Rxn<UserProfile>();
+  final mailboxDashBoardAction = Rxn<AppAction>();
```
Comment (hoangdat) [inline lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart:23]: I agree with the idea to classify the ViewEvent in online and offline, so please move this variable to `BaseController`.

Comment (hoangdat) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart:23]: AppAction -> ViewEvent

Comment (hoangdat) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart:23]: Besides that, we should not do it to broadcast the action from view to view, we should use viewState.
For this case:
```
1. open email -> request MarkAsRead
2. Update the state in EmailController then update the new state for MailboxDashboard also
3. All the subscriber of MailboxDashboardController will listen the event of its to react.
```

File: lib/features/email/presentation/email_controller.dart (hunk 3 above)
Comment (dab246) [reply lib/features/email/presentation/email_controller.dart:101]: What you say is quite confusing. Please describe in detail what you want.

Comment (hoangdat) [reply lib/features/email/presentation/email_controller.dart:101]: please read it here: https://github.com/linagora/tmail-flutter/pull/93#discussion_r709725474

File: lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart (hunk 4 above)
Comment (dab246) [reply lib/features/mailbox_dashboard/presentation/mailbox_dashboard_controller.dart:23]: Your scenario is similar to what I did. Please read it carefully.

**Review by hoangdat — APPROVED** [review]

**Pull request merged**